example/parallelLoopPatterns/src/parallelLoopPatterns.cpp

```diff
@@ -0,0 +1,452 @@
+              << elementsPerThread << " alpaka elements per thread\n";
+    alpaka::exec<TAcc>(queue, workDiv, ChunkedGridStridedLoopKernel{}, alpaka::getPtrNative(bufAcc), n);
+    testResult(queue, bufAcc);
+}
+
+//! A naive OpenMP style kernel mimicking a pragma omp parallel for loop with no chunk specified.
+struct NaiveOpenMPStyleKernel
+{
+    //! The kernel entry point.
+    //!
+    //! The work is distributed so that each thread processes a single consecutive range of elements.
+    //! The starting offset depends on global thread index and number of data elements.
+    //! For G = global thread index, T = global number of threads, N = number of data elements,
+    //! the mapping of thread index G to a set of processed data elements {D} is
+    //! G -> {D = G * ceil(N / T) + i, i = 0, 1, ..., ceil(N / T) - 1 | D < N}.
+    //! The same mapping is used for output buffer indices.
+    //!
+    //! This kernel can be run with any 1d work division.
+    //!
+    //! \tparam TAcc The accelerator environment to be executed on.
+    //! \param acc The accelerator to be executed on.
+    //! \param result The result array.
+    //! \param n The number of elements.
+    template<typename TAcc>
+    ALPAKA_FN_ACC void operator()(TAcc const& acc, float* result, uint32_t n) const
+    {
+        auto const globalThreadExtent(alpaka::getWorkDiv<alpaka::Grid, alpaka::Threads>(acc)[0u]);
+        auto const globalThreadIdx(alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0u]);
+        auto const processPerThread = (n + globalThreadExtent - 1) / globalThreadExtent;
+        for(uint32_t dataDomainIdx = globalThreadIdx * processPerThread;
+            (dataDomainIdx < (globalThreadIdx + 1) * processPerThread) && (dataDomainIdx < n);
+            dataDomainIdx++)
+        {
+            auto const memoryIdx = dataDomainIdx;
+            result[memoryIdx] = process(dataDomainIdx);
+        }
+    }
+};
+
+//! Data-parallel processing in naive OpenMP style:
+//! fixed number of threads and blocks, work per thread scales with problem size.
+//! This strategy can in principle be used with any alpaka accelerator, but numCores must be adjusted.
+//!
+//! \tparam TAcc The accelerator environment to be executed on.
+//! \tparam TQueue The queue type for work to be submitted to.
+//! \tparam TBufAcc The device buffer type.
+//! \param queue The queue object for work to be submitted to.
+//! \param bufAcc The device buffer.
+template<typename TAcc, typename TQueue, typename TBufAcc>
+void naiveOpenMPStyle(TQueue& queue, TBufAcc& bufAcc)
+{
+    auto const n = alpaka::getExtentProduct(bufAcc);
+    auto const deviceProperties = alpaka::getAccDevProps<TAcc>(alpaka::getDevByIdx<TAcc>(0u));
+    auto const maxThreadsPerBlock = deviceProperties.m_blockThreadExtentMax[0];
+    auto const numCores = 16u; // should be taken from hardware properties, alpaka currently does not expose it
```

Comment:
It is not great, but alpaka offers no way of doing it. I think we discussed that another library may do that. Note that it is not possible to take number of multiprocessors, as for CPU accelerators that would always be 1.

Reply:
What about
```c++
auto const numCores = std::max(std::thread::hardware_concurrency(), 1u);
```
?

Otherwise if you want more information, I'd suggest `hwloc` as a third part library (but only for convenience, since it is already an optional dependency of TBB, OpenMPI, _etc_).

Reply:
Thanks, I will try your std:: suggestion! For hwloc I wouldn't like to make a dependency in an example

Reply:
I included this suggestion in the PR.